sklearn/utils/_response.py

```diff
@@ -9,6 +9,110 @@
 from .validation import _check_response_method, check_is_fitted
 
 
+def _process_predict_proba(*, y_pred, target_type, classes, pos_label):
+    """Get the response values when the response method is `predict_proba`.
+
+    This function should process the `y_pred` in the binary and multi-label cases.
+    In the binary case, it should select the column corresponding to the positive
+    class. In the multi-label case, it should stack the predictions if they are not
+    in the "compressed" format `(n_samples, n_outputs)`.
+
+    Parameters
+    ----------
+    y_pred : ndarray
+        Output of `estimator.predict_proba`. The shape depends on the target type:
+
+        - for binary classification, it is a 2d array of shape `(n_samples, 2)`;
+        - for multiclass classification, it is a 2d array of shape
+          `(n_samples, n_classes)`;
+        - for multilabel classification, it is either a list of 2d arrays of shape
+          `(n_samples, 2)` (e.g. `RandomForestClassifier` or `KNeighborsClassifier`) or
+          an array of shape `(n_samples, n_outputs)` (e.g. `MLPClassifier` or
+          `RidgeClassifier`).
+
+    target_type : {"binary", "multiclass", "multilabel-indicator"}
+        Type of the target.
+
+    classes : ndarray of shape (n_classes,) or list of such arrays
+        Class labels as reported by `estimator.classes_`.
+
+    pos_label : int, float, bool or str
+        Only used with binary and multiclass targets.
+
+    Returns
+    -------
+    y_pred : ndarray of shape (n_samples,), (n_samples, n_classes) or \
+            (n_samples, n_output)
+        Compressed predictions format as requested by the metrics.
+    """
+    if target_type == "binary" and y_pred.shape[1] < 2:
+        # We don't handle classifiers trained on a single class.
+        raise ValueError(
+            f"Got predict_proba of shape {y_pred.shape}, but need "
+            "classifier with two classes."
+        )
+
+    if target_type == "binary":
+        col_idx = np.flatnonzero(classes == pos_label)[0]
+        return y_pred[:, col_idx]
+    elif target_type == "multilabel-indicator" and isinstance(y_pred, list):
+        # Use a compress format of shape `(n_samples, n_output)`.
+        # Only `MLPClassifier` and `RidgeClassifier` return an array of shape
+        # `(n_samples, n_outputs)`.
+        return np.vstack([p[:, -1] for p in y_pred]).T
+
+    return y_pred
+
+
+def _process_decision_function(*, y_pred, target_type, classes, pos_label):
+    """Get the response values when the response method is `decision_function`.
+
+    This function should process the `y_pred` in the binary and multi-label cases.
+    In the binary case, it should invert the sign of the score if the positive label
+    is not `classes[1]`. In the multi-label case, it should stack the predictions if
```

Comment:
this inversion of the sign is making me very uncomfortable, will need some time to see if it makes things explode in other places. Can you help me here?

Reply:
Actually, the sign inversion and selection selection of the columns is indeed part of bug fixes that we got since a couple of release now :).

You you change your `pos_label` then you flip the hyperplane. In terms of binary probability, it comes to take the other columns and therefore the `1 - y_proba`. Both, are consistent.